File: test/cpp/triangular_lattice.cc
#include "triangular_lattice.hh"
#include <boost/python.hpp>

// Function to be called from Python
void triangular_lattice_bind(GraphInterface& gi, int width, int height)
{
    // We need to invoke the appropriate instance of the template function
    // at runtime using the gt_dispatch<>() function, which handles the type
    // resolution. Here, we're considering all possible graph views and
    // the most basic vertex property maps (which we don't actually use here).
    gt_dispatch<>()(
        [&](auto& g) { create_triangular_lattice(g, width, height); },
        all_graph_views()
    )(gi.get_graph_view());
}

// Define the Python module that exposes the triangular lattice creation function
BOOST_PYTHON_MODULE(triangular_lattice)
{
    using namespace boost::python;
    def("create_triangular_lattice", triangular_lattice_bind);
}

File: src/lrgsglib/gt_patches/cpp/triangular_lattice.py
import graph_tool.all as gt

# Import the C++ module for creating a triangular lattice (called triangular_lattice.so)
import triangular_lattice

def create_triangular_lattice(width, height):
    """
    Create a triangular lattice graph with given width and height using the
    graph_tool library and a C++ extension for efficient implementation.

    Args:
        width (int): The width of the lattice.
        height (int): The height of the lattice.

    Returns:
        Graph: A graph_tool Graph object representing the triangular lattice.
    """
    # Create a new empty Graph using graph_tool
    g = gt.Graph(directed=False)

    # For graph objects, we need to pass the internal GraphInterface which is
    # accessed via g._Graph__graph.

    # The C++ function needs no property maps, so we only pass the graph and dimensions
    triangular_lattice.create_triangular_lattice(g._Graph__graph, width, height)

    return g
